cmd/eval_test.go

```diff
@@ -1042,3 +1067,34 @@ time.clock(input.y, time.clock(input.x))
 		})
 	}
 }
+
+func TestEvalWithStrictFlag(t *testing.T) {
+
+	input := `{
+		"foo": "2",
+		"b": [
+			{
+				"a": 1,
+				"b": [1, 2, 3],
+				"c": null
+			}
+		]
+	}`
+
+	query := `
+	package x
+
+	import future.keywords.if
+	import future.keywords.if
+	
+	foo = 2
+	
+
+	`
+
+	err := testEvalWithStrictFlag(input, query)
+	if err == nil {
+		t.Fatalf("expected error but err == nil")
+	}
+
```

Comment:
This test case needs a bit more work:

1. It's got unrelated stuff in it: `input` isn't relevant to the policy at all
2. There' no need to split off `testEvalWithStrictFlag` when it's not used multiple times, it can be inlined here
3. The error is what we care about, so let's check more than just if it is `!= nil`
4. A few more cases would be nice: especially, it's desired that the strictness applies to the _policy_ only, not the _query_. So if the query is `x := data.x.foo`, that `x` being _unused_ should not be a problem. It would be nice if the test cases reflected that.

Reply:
Thanks! Apologies for being a bit of a pain with this I kind-of had an idea but needed some guidance. Will work on this!

Reply:
Just wanted to keep you posted, here is my second iteration on the testing its a bit more fleshed out but not sure if its a bit over the top or "over engineered" for what i needs to do:

```
func TestWithStrictFlag(t *testing.T) {
	testsShouldError := []struct {
		policy          string
		query           string
		expectedCode    string
		expectedMessage string
	}{
		{
			policy:          "package x\n import future.keywords.if\n import future.keywords.if\n foo = 2",
			query:           "data.foo",
			expectedCode:    "rego_compile_error",
			expectedMessage: "import must not shadow import future.keywords.if",
		},
		{
			policy:          "package x\nimport future.keywords.if\n import data.foo\n foo = 2",
			query:           "data.foo",
			expectedCode:    "rego_compile_error",
			expectedMessage: "import data.foo unused",
		},
		{
			policy:          "package x\nimport future.keywords.if\n data if { x = 1}",
			query:           "data.foo",
			expectedCode:    "rego_compile_error",
			expectedMessage: "rules must not shadow data (use a different rule name)",
		},
	}

	for _, tc := range testsShouldError {
		t.Run("These Tests Should Return Errors", func(t *testing.T) {

			files := map[string]string{
				"test.rego": tc.policy,
			}

			test.WithTempFS(files, func(path string) {
				params := newEvalCommandParams()
				params.strict = true

				_ = params.dataPaths.Set(filepath.Join(path, "test.rego"))

				var buf bytes.Buffer
				_, err := eval([]string{tc.query}, params, &buf)
				if err == nil {
					t.Fatal("expected error, got nil")
				}
				var output presentation.Output
				if err := util.NewJSONDecoder(&buf).Decode(&output); err != nil {
					t.Fatal(err)
				}

				if code := output.Errors[0].Code; code != tc.expectedCode {
					t.Errorf("expected code '%v', got '%v'", tc.expectedCode, code)
				}
				if msg := output.Errors[0].Message; msg != tc.expectedMessage {
					t.Errorf("expected message '%v', got '%v'", tc.expectedMessage, msg)
				}
			})
		})
	}

	testsShouldPass := []struct {
		policy string
		query  string
	}{
		{
			policy: "package x\n import future.keywords.if\n foo = 2",
			query:  "data.foo",
		},
		{
			policy: "package x\n import future.keywords.if\n foo = 2",
			query:  "x := data.foo",
		},
	}
	for _, tc := range testsShouldPass {
		t.Run("These Tests Should Return Errors", func(t *testing.T) {

			files := map[string]string{
				"test.rego": tc.policy,
			}

			test.WithTempFS(files, func(path string) {
				params := newEvalCommandParams()
				params.strict = true

				_ = params.dataPaths.Set(filepath.Join(path, "test.rego"))

				var buf bytes.Buffer
				_, err := eval([]string{tc.query}, params, &buf)
				if err != nil {
					t.Fatal("Should not error, got error:", err)
				}
			})
		})
	}

}
```

Not fully finished but I think this is much better than before? :smiling_face_with_tear:

Reply:
Doesn't look bad at all. The only thing standing out to me is this:

```
		t.Run("These Tests Should Return Errors", func(t *testing.T) {
```
let's rather a short note to each test case and do `r.Run(tc.note, func(t *testing.T) { ... })` instead.